github/compliance/issue.py

```diff
@@ -192,42 +164,39 @@ def _update_issue(
 
 
 def create_or_update_issue(
-    component: cm.Component,
-    artifact: cm.Artifact,
+    labels: frozenset[str],
     repository: github3.repos.Repository,
-    body:str,
+    body: str,
     known_issues: typing.Iterable[github3.issues.issue.ShortIssue],
-    title:str=None,
+    title: str,
     assignees: typing.Iterable[str]=(),
     milestone: github3.issues.milestone.Milestone=None,
-    issue_type: str=_label_bdba,
     extra_labels: typing.Iterable[str]=None,
     preserve_labels_regexes: typing.Iterable[str]=(),
 ) -> github3.issues.issue.ShortIssue:
     open_issues = tuple(
         enumerate_issues(
-            component=component,
-            artifact=artifact,
+            labels=labels,
             known_issues=known_issues,
-            issue_type=issue_type,
             state='open',
         )
     )
     if (issues_count := len(open_issues)) > 1:
-        raise RuntimeError(f'more than one open issue found for {component.name=}{artifact.name=}')
+        raise RuntimeError(f'more than one open issue found for {labels=}')
```

Comment:
hmm, that's not nice - those labels are quite cumbersome to read, considering the component/artefact-names are only there as hash-digests

Thus, I think it should be better to change the passed-in "target-identifier", like so:
- instead of component and resource, use `cnudie.iter.Node` (or for a first iteration, we might only accept ResourceNode)
- for cfg-elements, we might accept cfg-element

I am considering something like so:

```
def enumerate_issues(
 target: cnudie.iter.ResourceNode | model.base.NamedModelElement
 .. <other parameters, as they were defined before>
):
 labels = labels_from_target(target=target)
```

def labels_from_target(target: cnudie.iter.ResourceNode | model.base.NamedModelElement):
 if isinstance(target, RN):
   <code from before this change>
 elif isinstance(NME):
  <read element's name and typename to create labels>
 else:
  raise NotImplementedError(target)
```

Reply:
We have to at least implement `cnudie.iter.ResourceNode` and `cnudie.iter.SourceNode` targets as checkmarx scans are in place already.

How about introducing a dataclass?

```python
@dc
class Target:
    target: ResourceNode | SourceNode | ...

    # could be used for logging
    @property
    def name(self) -> str:
        if isinstance(cnudie.iter.ResourceNode):
            return f'{self.target.component.name}:{self.target.resource.name}'

        if isinstance(cnudie.iter.SourceNode):
            return f'{self.target.component.name}:{self.target.source.name}'

    # determining target digest label here, so repository_labels could take an optional
    # digest_label callback and yield "common" labels (cicd/auto-generated, ...)
    def digest_label(self) -> str:
        if isinstance(self.target, cnudie.iter.ResourceNode):
            return artifact_digest_label(
                component=self.target.component,
                artifact=self.target.resource,
            )
        if isinstance(self.target, cnudie.iter.SourceNode):
            return artifact_digest_label(
                component=self.target.component,
                artifact=self.target.source,
            )
        # can be extended for efg-element
```

Reply:
well, let's support ResourceNode and SourceNode in parallel, then - why introduce a new class for model elements that already have one?

Reply:
> why introduce a new class for model elements that already have one?

to group functions for "targets" like `name()` or `digest_label()`.

Reply:
I consider this avoidable boilerplate.

just define a function that accepts a target and returns labels for the target, and - if required multiple times - a second function returning the display-name.

the structure as suggested above mixes different concerns, for which even dedicated functions already exist.

if at all, in order to reduce the amount of functions that need to dispatch between different target-types (from two to one), one could consider to define a dataclass holding name and labels. tbh I do not see much value here, and such an approach will not lead to a better codebase